Incorrect comment assigned to function expression

With the change in 0d50df705f1df2d1486c7fdad14ab0aefd31215f, an incorrect comment block may be assigned to a function expression.

This used to pass `valid-jsdoc`, now it generates `Missing JSDoc @returns for function`:

```js
/* eslint valid-jsdoc: [2, {requireReturn: false}] */

/**
 * Create a new thing.
 */
var thing = new Thing({
  foo: function() {
    return 'bar';
  }
});
```


<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29736869-incorrect-comment-assigned-to-function-expression?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

> 1. The version of ESLint you are using (run eslint -v)

This looks like a regression introduced by 0d50df705f1df2d1486c7fdad14ab0aefd31215f

> 2. What you did (the source code and ESLint configuration)

Here is the source code and complete config (same as in the original description):
```
/* eslint valid-jsdoc: [2, {requireReturn: false}] */

/**
 * Create a new thing.
 */
var thing = new Thing({
  foo: function() {
    return 'bar';
  }
});
```

> 3. The actual ESLint output complete with numbers

```
$ git checkout 0d50df705f1df2d1486c7fdad14ab0aefd31215f
$ node bin/eslint.js --no-eslintrc script.js

/Users/tschaub/projects/eslint/script.js
  3:1  error  Missing JSDoc @returns for function  
```

> 4. What you expected to happen instead

I would expect the above code to be valid according to `valid-jsdoc`.  The documentation block is not associated with the function expression.

Prior to 0d50df705f1df2d1486c7fdad14ab0aefd31215f, the `valid-jsdoc` rule treated this code as valid (as it should):

```
$ git checkout 0d50df705f1df2d1486c7fdad14ab0aefd31215f^
$ node bin/eslint.js --no-eslintrc script.js
$
```

The issue is that the `sourceCode.getJSDocComment()` traverses up to the assignment expression, finds the leading comments, and returns these as the comment associated with the function expression.

--- a/lib/util/source-code.js
+++ b/lib/util/source-code.js
@@ -233,7 +233,7 @@ SourceCode.prototype = {
             case "FunctionExpression":
 
                 if (parent.type !== "CallExpression" && parent.type !== "NewExpression") {
-                    while (parent && !parent.leadingComments && !/Function/.test(parent.type) && parent.type !== "MethodDefinition") {
+                    while (parent && !parent.leadingComments && !/Function/.test(parent.type) && parent.type !== "MethodDefinition" && parent.type !== "Property") {
                         parent = parent.parent;
                     }
 

--- a/tests/lib/rules/valid-jsdoc.js
+++ b/tests/lib/rules/valid-jsdoc.js
@@ -40,6 +40,33 @@ ruleTester.run("valid-jsdoc", rule, {
         "/**\n* Description\n* @inheritdoc */\nfunction foo(arg1, arg2){ return ''; }",
         "/**\n* Description\n* @inheritDoc */\nfunction foo(arg1, arg2){ return ''; }",
         {
+            code:
+                "/**\n" +
+                "* Create a new thing.\n" +
+                "*/\n" +
+                "var thing = new Thing({\n" +
+                "  foo: function() {\n" +
+                "    return 'bar';\n" +
+                "  }\n" +
+                "});\n",
+            options: [{requireReturn: false}]
+        },
+        {
+            code:
+                "/**\n" +
+                "* Create a new thing.\n" +
+                "*/\n" +
+                "var thing = new Thing({\n" +
+                "  /**\n" +
+                "   * @return {string} A string.\n" +
+                "   */\n" +
+                "  foo: function() {\n" +
+                "    return 'bar';\n" +
+                "  }\n" +
+                "});\n",
+            options: [{requireReturn: false}]
+        },
+        {
             code: "/**\n* Description\n* @return {void} */\nfunction foo(){}",
             options: [{}]
         },
@@ -231,6 +258,25 @@ ruleTester.run("valid-jsdoc", rule, {
             code: "/** @@foo */\nfunction foo(){}",
             errors: [{
                 message: "JSDoc syntax error.",
+                type: "Block"
+            }]
+        },
+        {
+            code:
+                "/**\n" +
+                "* Create a new thing.\n" +
+                "*/\n" +
+                "var thing = new Thing({\n" +
+                "  /**\n" +
+                "   * Missing return tag.\n" +
+                "   */\n" +
+                "  foo: function() {\n" +
+                "    return 'bar';\n" +
+                "  }\n" +
+                "});\n",
+            options: [{requireReturn: false}],
+            errors: [{
+                message: "Missing JSDoc @returns for function.",
                 type: "Block"
             }]
         },

